irods/test/access_test.py

```diff
@@ -371,6 +372,81 @@ def test_removed_user_does_not_affect_raw_ACL_queries__issue_431(self):
+
+        finally:
+            self.sess.users.get(user_name).remove()
+            self.sess.collections.remove(collection_path, force=True)
+
+    def test_iRODSAccess_can_be_constructed_using_iRODSDataObject__issue_558(self):
+        user_name = "testuser"
+        data_object_path = "/".join([helpers.home_collection(self.sess), "give_read_access_to_this"])
+
+        try:
+            user = self.sess.users.create(user_name, 'rodsuser')
+            data_object = self.sess.data_objects.create(data_object_path)
+
+            # Give user access to data object. This should succeed. Before the fix in #558, this would cause an error
+            # from pickle during a call to deepcopy of the iRODSAccess object. The library does not know how to pickle
+            # an iRODSDataObject object.
+            access = iRODSAccess('read', data_object, user.name)
+            self.sess.acls.set(access)
+
+            # We can get permissions from the data object, and the test user's entry is there.
+            perms = self.sess.acls.get(data_object)
+            self.assertTrue(any(p for p in perms if p.user_name == user_name))
+
+        finally:
+            self.sess.users.get(user_name).remove()
+            self.sess.data_objects.unlink(data_object_path, force=True)
+
+    def test_iRODSAccess_can_be_constructed_using_iRODSPath__issue_558(self):
+        user_name = "testuser"
+        data_object_path = "/".join([helpers.home_collection(self.sess), "give_read_access_to_this"])
+
+        try:
+            irods_path = iRODSPath(data_object_path)
+
+            user = self.sess.users.create(user_name, 'rodsuser')
+            data_object = self.sess.data_objects.create(data_object_path)
+
+            # Give user access to data object. This should succeed. Before the fix in #558, this would cause an error
+            # from pickle during a call to deepcopy of the iRODSAccess object. The library does not know how to pickle
+            # an iRODSDataObject object.
+            access = iRODSAccess('read', irods_path, user.name)
+            self.sess.acls.set(access)
+
+            # We can get permissions from the data object, and the test user's entry is there.
+            perms = self.sess.acls.get(data_object)
+            self.assertTrue(any(p for p in perms if p.user_name == user_name))
+
+        finally:
+            self.sess.users.get(user_name).remove()
+            self.sess.data_objects.unlink(data_object_path, force=True)
+
+    def test_iRODSAccess_cannot_be_constructed_using_unsupported_type__issue_558(self):
+        # Before the fix in #558, this would have been allowed and only later would the type discrepancy be revealed,
+        # leading to opaque error messages. Now, the types are checked on the way in to ensure clarity and correctness.
+        with self.assertRaises(TypeError):
+            _ = iRODSAccess('read', self.sess, self.sess.username)
```

Comment:
Is it possible to assert **part** of the error message to guard against something else causing a `TypeError`?

Reply:
I just put the whole error message because it is static. I also noticed that we cannot use the context manager here since that was introduced in python 3.1. We can rework this as part of the effort for #480.